src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/AndroidX509CertificateReader.cs

```diff
@@ -0,0 +1,492 @@
+        private X500DistinguishedName? _issuerName;
+        private string? _subject;
+        private string? _issuer;
+
+        public static ICertificatePal FromHandle(IntPtr handle)
+        {
+            if (handle == IntPtr.Zero)
+                throw new ArgumentException(SR.Arg_InvalidHandle, nameof(handle));
+
+            return new OpenSslX509CertificateReader(Interop.Crypto.X509UpRef(handle));
+        }
+
+        public static ICertificatePal FromOtherCert(X509Certificate cert)
+        {
+            Debug.Assert(cert.Pal != null);
+
+            // Ensure private key is copied
+            OpenSslX509CertificateReader certPal = (OpenSslX509CertificateReader)cert.Pal;
+            return certPal.DuplicateHandles();
+        }
+
+        public static ICertificatePal FromBlob(ReadOnlySpan<byte> rawData, SafePasswordHandle password, X509KeyStorageFlags keyStorageFlags)
+        {
+            Debug.Assert(password != null);
+            ICertificatePal? cert;
+            Exception? exception;
+            if (TryReadX509(rawData, out cert)
+                || PkcsFormatReader.TryReadPkcs7Der(rawData, out cert)
+                || PkcsFormatReader.TryReadPkcs7Pem(rawData, out cert)
+                || PkcsFormatReader.TryReadPkcs12(rawData, password, out cert, out exception))
+            {
+                if (cert == null)
+                {
+                    // Empty collection, most likely.
+                    throw new CryptographicException();
+                }
+
+                return cert;
+            }
+
+            // Unsupported
+            Debug.Assert(exception != null);
+            throw exception;
+        }
+
+        public static ICertificatePal FromFile(string fileName, SafePasswordHandle password, X509KeyStorageFlags keyStorageFlags)
+        {
+            byte[] fileBytes = System.IO.File.ReadAllBytes(fileName);
+            return FromBlob(fileBytes, password, keyStorageFlags);
+        }
+
+        // Handles both DER and PEM
+        internal static bool TryReadX509(ReadOnlySpan<byte> rawData, [NotNullWhen(true)] out ICertificatePal? handle)
+        {
+            handle = null;
+            SafeX509Handle certHandle = Interop.AndroidCrypto.DecodeX509(
+                ref MemoryMarshal.GetReference(rawData),
+                rawData.Length);
+
+            if (certHandle.IsInvalid)
+            {
+                certHandle.Dispose();
+                return false;
+            }
+
+            handle = new OpenSslX509CertificateReader(certHandle);
+            return true;
+        }
+
+        internal static bool TryReadX509Der(ReadOnlySpan<byte> rawData, [NotNullWhen(true)] out ICertificatePal? certPal)
+        {
+            return TryReadX509(rawData, out certPal);
+        }
+
+        internal static bool TryReadX509Pem(ReadOnlySpan<byte> rawData, [NotNullWhen(true)] out ICertificatePal? certPal)
+        {
+            return TryReadX509(rawData, out certPal);
+        }
```

Comment:
Hopefully (a) this gets cleaned up as less code is shared and (b) there aren't callers that are assuming that the Der version rejects PEM and that the Pem version rejects DER.